### PR TITLE
Failed message retry

### PR DESCRIPTION
**Pull request by tompro**
## 📝 Description

Adds retry job for sending chain events. Messages will be retried up to 10 times triggered by the current scheduler. If we want to get a bit more efficient we can add an exponential back-off later.

Relates to #297

---

## ✅ Checklist

Please ensure the following tasks are completed before requesting a review:

- [x] My code adheres to the coding guidelines of this project.
- [x] I have run `cargo fmt`.
- [x] I have run `cargo clippy`.
- [x] I have added or updated tests (if applicable).
- [x] All CI/CD steps were successful.
- [x] I have updated the documentation (if applicable).
- [x] I have checked that there are no console errors or warnings.
- [x] I have verified that the application builds without errors.
- [x] I've described the changes made to the API. (modification, addition, deletion).

---

## 🚀 Changes Made

- **New Features:**
  - Automatically retry sending Nostr chain events if delivery fails

---

## 💡 How to Test

Please provide clear instructions on how reviewers can test your changes:

1. cargo test

---

## 🤝 Related Issues

List any related issues, pull requests, or discussions:

- #424

---

## 📋 Review Guidelines

Please focus on the following while reviewing:

- [ ] Does the code follow the repository's contribution guidelines?
- [ ] Are there any potential bugs or performance issues?
- [ ] Are there any typos or grammatical errors in the code or comments?

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced a retry mechanism for failed Nostr message deliveries.
  - Added a queue to store failed messages with retry logic.
  - Implemented a job to process and retry queued messages.

- Enhanced `DbContext` to include `NostrQueuedMessageStoreApi`.

- Updated notification service to handle retries and integrate with the message queue.

- Added comprehensive tests for retry logic, including edge cases and error handling.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>7 files</summary><table>
<tr>
  <td><strong>mod.rs</strong><dd><code>Added NostrQueuedMessageStoreApi to DbContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-d89d28d4200c90484ca080e04499928227cf65ee7857931621617d034845ea9c">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>default_service.rs</strong><dd><code>Implemented retry logic for failed Nostr messages</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-a7465b19a0de744da31082e747b854d7196296da443f668f516f58e7936423bf">+546/-18</a></td>

</tr>

<tr>
  <td><strong>mod.rs</strong><dd><code>Updated notification service creation to include retry queue</code></dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-5e4582d88d28c90be7fe309bfd3facef93108f563c5cb55fa44350316e27a5e6">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>notification_service.rs</strong><dd><code>Extended NotificationServiceApi with retry functionality</code>&nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-3d0a1f93baac17cad711250480de3ed125953511c77cb770e2b5ba81274e18f1">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>context.rs</strong><dd><code>Integrated retry queue into WASM context</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-4aab47a7683b01dc13d3bd22573728e6f581df241ee8a3e6eb19d12521bd635f">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>job.rs</strong><dd><code>Added job to process Nostr message retry queue</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-739e234452ac68eb6d4ee619fed093bc9fffcff992ef48215f9ee60fd05de37a">+10/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>service_context.rs</strong><dd><code>Integrated retry queue into web service context</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-2bc9ef2eca1470d42d8c155ab475fda277a11bfefee32a9a6f0a48f447169b0e">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>test_utils.rs</strong><dd><code>Added mock for NostrQueuedMessageStore</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-cf644a00547031931e4687af48d33baf744a1a001e5b215caa46753fd43dff39">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mod.rs</strong><dd><code>Added tests for retry logic and NostrQueuedMessageStore</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/BitcreditProtocol/E-Bill/pull/448/files#diff-49db059de3206a221445a6f9d2745edf8e81decc03c3ac08e49ceb1d727b2a02">+14/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: crates/bcr-ebill-api/src/service/notification_service/default_service.rs [hunk 1]
```diff
@@ -368,6 +399,42 @@ impl NotificationServiceApi for DefaultNotificationService {
...
+
+    async fn send_retry_messages(&self) -> Result<()> {
+        let mut failed_ids = vec![];
+        while let Ok(Some(queued_message)) = self
+            .queued_message_store
+            .get_retry_messages(1)
+            .await
+            .map(|r| r.first().cloned())
+        {
+            println!("got next retry message {q}", q = queued_message.id);
```
Comment (zupzup) [inline crates/bcr-ebill-api/src/service/notification_service/default_service.rs:411]: leftover println - could maybe be a `debug!` log?

**Review by zupzup — APPROVED** [review]
LGTM :+1:
